### PR TITLE
Contract call with tuple is missing param names

### PR DESCRIPTION
**Pull request by avkos**
## Description

#5504 

## Checklist for 4.x:

- [ ] I have selected the correct 4.x base branch.
- [ ] Within the description, the feature or issue is discussed in detail or an issue is linked.   
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have added any required tests for the changes I made 
- [ ] I have made corresponding changes to the documentation
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] I ran `yarn` successfully
- [ ] I ran `yarn lint` successfully
- [ ] I ran `yarn build:web` successfully
- [ ] I ran `yarn test:unit` successfully
- [ ] I ran `yarn test:integration` successfully
- [ ] I ran `compile:contracts` successfully
- [ ] I have tested my code.
- [ ] I have updated the corresponding `CHANGELOG.md` file in the packages I have edited.

### CODE REVIEW

File: packages/web3-eth-contract/src/types.ts [hunk 1]
```diff
@@ -227,7 +227,11 @@ export interface NonPayableMethodObject<Inputs = unknown[], Outputs = unknown[]>
 	 * @param block - If you pass this parameter it will not use the default block set with contract.defaultBlock. Pre-defined block numbers as `earliest`, `latest`, and `pending` can also be used. Useful for requesting data from or replaying transactions in past blocks.
 	 * @returns - The return value(s) of the smart contract method. If it returns a single value, it’s returned as is. If it has multiple return values they are returned as an object with properties and indices.
 	 */
-	call(tx?: NonPayableCallOptions, block?: BlockNumberOrTag): Promise<Outputs>;
+
+	call<SpecialOutput = undefined>(
```
Comment (mpetrunic) [inline packages/web3-eth-contract/src/types.ts]: why not just default to Outputs?

**Review by mpetrunic — COMMENTED** [review]
This works sort of, I was hoping you would solve this: https://github.com/web3/web3.js/blob/4.x/packages/web3-eth-abi/src/types.ts#L193 😜

**Comment by avkos** [comment]
> This works sort of, I was hoping you would solve this: https://github.com/web3/web3.js/blob/4.x/packages/web3-eth-abi/src/types.ts#L193 😜

I've tried to figure it out but seems typescript not supporting dynamic names for the tuples right now.

File: packages/web3-eth-contract/src/types.ts (hunk 1 above)
Comment (avkos) [reply packages/web3-eth-contract/src/types.ts]: because later I use conditional type and check if it is undefined then use Outputs

Comment (mpetrunic) [reply packages/web3-eth-contract/src/types.ts]: I understand, but wouldn't my suggestion be simpler? 😄

Comment (avkos) [reply packages/web3-eth-contract/src/types.ts]: fixed. thanks

**Review by Muhammad-Altabba — DISMISSED** [review]

**Comment by mpetrunic** [comment]
@avkos Here is an idea https://github.com/web3/web3.js/pull/5624 ^^

**Review by nikoulai — APPROVED** [review]

**Comment by avkos** [comment]
> @avkos Here is an idea #5624 ^^

Thanks a lot. I'll investigate it.

**Review by jdevcs — COMMENTED** [review]
Its causing many tests to fail in https://github.com/web3/web3.js/actions/runs/3490219406/jobs/5841458348#step:4:176 all these tests should be passing before merging PR.

**Comment by mpetrunic** [comment]
> Its causing many tests to fail in https://github.com/web3/web3.js/actions/runs/3490219406/jobs/5841458348#step:4:176 all these tests should be passing before merging PR.

Even though types should be fixed to support that, it makes more sense to use output param name instead of array index in those tests as it's much more type safer.

It might make sense to introduce type only tests for web3-eth-abi package: https://medium.com/front-end-weekly/testing-typescript-types-1595c641d7c7^^

**Comment by mpetrunic** [comment]
@avkos https://github.com/web3/web3.js/pull/5647 adds type tests and support for single output tuple

**Comment by avkos** [comment]
-  add the possibility to see types in the output types

image

- if single output returns exactly one type
image

- for overloaded functions shows all variants of input parameters
image

**Comment by mpetrunic** [comment]
> * add the possibility to see types in the output types
> 
> image
> 
> * if single output returns exactly one type
> 
> image
> 
> * for overloaded functions shows all variants of input parameters
> 
> image

Think you should check my PR as I've extended tests so you can cover all this cases with tests to make sure everything works^^

**Comment by avkos** [comment]
I've fixed the types. Now they work in both ways by index and by name. Also, type tests were added. @mpetrunic

**Comment by mpetrunic** [comment]
> I've fixed the types. Now they work in both ways by index and by name. Also, type tests were added. @mpetrunic

@jdevcs @mconnelly8 Would it make sense to open an issue to create a guide for users explaining how to extract result types from abi (as they will probably need to do something with the result)? Like `type GetBalanceResult = SomethingSomething<typeof abi[1]>`

**Review by Spacesai1or — CHANGES_REQUESTED** [review]
Please add `CHANGELOG.md` entries (for root, `web3-eth-abi`, and `web3-eth-contract`)

**Comment by avkos** [comment]
> Please add `CHANGELOG.md` entries (for root, `web3-eth-abi`, and `web3-eth-contract`)

Changelogs were added

**Review by Spacesai1or — APPROVED** [review]

**Pull request merged**